client/findproviders.go

```diff
@@ -23,18 +25,34 @@ type DelegatedRoutingClient interface {
 	GetIPNSAsync(ctx context.Context, id []byte) (<-chan GetIPNSAsyncResult, error)
 	PutIPNS(ctx context.Context, id []byte, record []byte) error
 	PutIPNSAsync(ctx context.Context, id []byte, record []byte) (<-chan PutIPNSAsyncResult, error)
-	Provide(ctx context.Context, key cid.Cid, provider peer.AddrInfo, ttl time.Duration) (<-chan time.Duration, error)
+	Provide(ctx context.Context, key cid.Cid, ttl time.Duration) (time.Duration, error)
+	ProvideAsync(ctx context.Context, key cid.Cid, ttl time.Duration) (<-chan time.Duration, error)
 }
 
 type Client struct {
 	client    proto.DelegatedRouting_Client
 	validator record.Validator
+
+	provider *Provider
+	identity crypto.PrivKey
 }
 
+var _ DelegatedRoutingClient = (*Client)(nil)
+
 func NewClient(c proto.DelegatedRouting_Client) *Client {
 	return &Client{client: c, validator: ipns.Validator{}}
 }
 
+// SetIdentity sets an identity for providing content. the `Provide` methods will not work without it set.
+func (c *Client) SetIdentity(p *Provider, identity crypto.PrivKey) error {
```

Comment:
Should we set identity as a constructor param? if it is mandatory, I think it should be set on the constructor.

Reply:
you can use the other find / ipns methods without an identity - so it's only needed for calling the provide method. I left it as optional because of that. what do you think?

Reply:
I would say to specify that case on the constructor documentation and make it possible for the caller to set `Provider` and identity as nil. I think that way of implementing it will cause less confusion and runtime errors for implementors.